graph/path/path.go

```diff
@@ -0,0 +1,296 @@
+// Copyright 2014 The Cayley Authors. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package path
+
+import (
+	"github.com/google/cayley/graph"
+	"github.com/google/cayley/graph/iterator"
+	"github.com/google/cayley/quad"
+)
+
+type morphism struct {
+	Name     string
+	Reversal func() morphism
+	Apply    graph.ApplyMorphismFunc
+}
+
+type Path struct {
+	stack []morphism
+	qs    graph.QuadStore // Optionally. A nil qs is equivalent to a morphism.
+}
+
+func (p *Path) IsMorphism() bool { return p.qs == nil }
+
+func StartMorphism(nodes ...string) *Path {
+	return StartPath(nil, nodes...)
+}
+
+func StartPath(qs graph.QuadStore, nodes ...string) *Path {
+	return &Path{
+		stack: []morphism{
+			isMorphism(nodes...),
+		},
+		qs: qs,
+	}
+}
+
+func PathFromIterator(qs graph.QuadStore, it graph.Iterator) *Path {
+	return &Path{
+		stack: []morphism{
+			iteratorMorphism(it),
+		},
+		qs: qs,
+	}
+}
+
+func NewPath(qs graph.QuadStore) *Path {
+	return &Path{
+		qs: qs,
+	}
+}
+
+func (p *Path) Reverse() *Path {
+	newPath := NewPath(p.qs)
+	for i := len(p.stack) - 1; i >= 0; i-- {
+		newPath.stack = append(newPath.stack, p.stack[i].Reversal())
+	}
+	return newPath
+}
+
+func (p *Path) Is(nodes ...string) *Path {
+	p.stack = append(p.stack, isMorphism(nodes...))
+	return p
+}
+
+func (p *Path) Tag(tags ...string) *Path {
+	p.stack = append(p.stack, tagMorphism(tags...))
+	return p
+}
+
+func (p *Path) Out(via ...interface{}) *Path {
+	p.stack = append(p.stack, outMorphism(via...))
+	return p
+}
+func (p *Path) In(via ...interface{}) *Path {
+	p.stack = append(p.stack, inMorphism(via...))
+	return p
+}
+
+func (p *Path) And(path *Path) *Path {
+	p.stack = append(p.stack, andMorphism(path))
+	return p
+}
+
+func (p *Path) Or(path *Path) *Path {
+	p.stack = append(p.stack, orMorphism(path))
+	return p
+}
+
+func (p *Path) Except(path *Path) *Path {
+	p.stack = append(p.stack, exceptMorphism(path))
+	return p
+}
+
+func (p *Path) Follow(path *Path) *Path {
+	p.stack = append(p.stack, followMorphism(path))
+	return p
+}
+
+func (p *Path) FollowReverse(path *Path) *Path {
+	p.stack = append(p.stack, followMorphism(path.Reverse()))
+	return p
+}
+
+func (p *Path) BuildIterator() graph.Iterator {
```

Comment:
My thinking exactly. A program which executes such a query (namely, one that is malformed) should panic. The query may be *wrong* but at least it's *valid*. ;)